components/engine_traits/src/engine.rs

```diff
@@ -64,6 +64,14 @@ pub trait KvEngine:
     /// This only exists as a temporary hack during refactoring.
     /// It cannot be used forever.
     fn bad_downcast<T: 'static>(&self) -> &T;
+
+    /// Returns false if KvEngine can't apply snapshot for this region now.
+    /// Some KvEngines need to do some transforms before apply data from
+    /// snapshot. These procedures can be batched in background if there are
+    /// more than one incoming snapshots, thus not blocking applying thread.
+    fn can_apply_snapshot(&self, _is_timeout: bool, _region_id: u64) -> bool {
```

Comment:
Why need `is_timeout`? How can it make any difference?

Reply:
If there is no incoming applying snapshot for a long time, `handle_pending_applies` can be triggered by `on_timeout`, thus proxy can do it directly in foreground.

Reply:
Then why bother calling `can_apply_snapshot` again?

Reply:
`handle_pending_applies` will apply multiple snapshots as long as it will not cause write stall. However, in proxy we won't handle the more than one snapshot if the following snapshot is still pre-handling.